stable_baselines3/dqn/dqn.py

```diff
@@ -0,0 +1,210 @@
+        self.exploration_initial_eps = exploration_initial_eps
+        self.exploration_fraction = exploration_fraction
+        self.target_update_interval = target_update_interval
+        self.tau = tau
+        self.max_grad_norm = max_grad_norm
+        self.tensorboard_log = tensorboard_log
+
+        if _init_setup_model:
+            self._setup_model()
+
+    def _setup_model(self) -> None:
+        super(DQN, self)._setup_model()
+        self._create_aliases()
+        self._setup_exploration_schedule()
+
+    def _setup_exploration_schedule(self) -> None:
+        """
+        Generate a exploration schedule used for updating the exploration probability
+        """
+        self.exploration_schedule = get_linear_fn(self.exploration_initial_eps, self.exploration_final_eps,
+                                                  self.exploration_fraction)
+
+    def _update_exploration(self) -> None:
+        """
+        Update the policy exploration probability using the current exploration rate schedule
+        and the current progress (from 1 to 0).
+        """
+        # Log the current exploration probability
+        logger.logkv("exploration rate", self.exploration_schedule(self._current_progress))
+
+        self.policy.update_exploration_rate(self.exploration_schedule(self._current_progress))
+
+    def _create_aliases(self) -> None:
+        self.q_net = self.policy.q_net
+        self.q_net_target = self.policy.q_net_target
+
+    def train(self, gradient_steps: int, batch_size: int = 100) -> None:
+        # Update learning rate and exploration probability according to schedule
+        self._update_learning_rate(self.policy.optimizer)
+        self._update_exploration()
+
+        for gradient_step in range(gradient_steps):
+            # Sample replay buffer
+            replay_data = self.replay_buffer.sample(batch_size, env=self._vec_normalize_env)
+
+            with th.no_grad():
+                # Compute the target Q values
+                target_q = self.q_net_target(replay_data.next_observations)
+                target_q, _ = target_q.max(dim=1)
+                target_q = target_q.reshape(-1, 1)
+                target_q = replay_data.rewards + (1 - replay_data.dones) * self.gamma * target_q
+
+            # Get current Q estimates
+            current_q = self.q_net(replay_data.observations)
+
+            # Gather q_values of our actions
+            current_q = th.gather(current_q, dim=1, index=replay_data.actions.long())
+
+            # Compute q loss
+            loss = F.mse_loss(current_q, target_q)
+
+            # Optimize the policy
+            self.policy.optimizer.zero_grad()
+            loss.backward()
+            # Clip grad norm
+            th.nn.utils.clip_grad_norm_(self.policy.parameters(), self.max_grad_norm)
+            self.policy.optimizer.step()
+
+            # Update target networks
+            if self._n_updates % self.target_update_interval == 0:
```

Comment:
you should update `self._n_updates` before, otherwise this condition may never be true